File: BOJ/12000-12999/12178.cpp
#include <iostream>
#include <cmath>

using namespace std;

int multifactorial[9000];

int calc_multifactorial(int n) {
    double ret = 0;
    
    for (int i=9000; i>=1; i-=n)
        ret += log10(i);
    
    return (int)ret + 1;
}

void fill_multifactorial() {
    for (int i=1; i<9000; i++) {
        multifactorial[i] = calc_multifactorial(i);
    }
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);
    
    fill_multifactorial();
    
    int T;
    cin >> T;
    
    for (int t=1; t<=T; t++) {
        int D;
        cin >> D;
        
        int ans = 0;
        
        for (int i=1; i<9000; i++) {
            if (D > multifactorial[i]) {
                ans = i;
                break;
            }
        }
        
        cout << "Case #" << t << ": ";
        
        if (ans == 0) cout << "...";
        else {
            cout << "IT'S OVER 9000";
            for (int i=0; i<ans; i++) cout << "!";
        }
        
        cout << "\n";
    }

    return 0;
}

File: BOJ/12000-12999/12178.py
T = int(input())
multifactorial = []

def lower_bound(target):
    lo = 0
    hi = len(multifactorial)
    
    while lo < hi:
        mid = (lo + hi) // 2
        
        if multifactorial[mid] < target:
            lo = mid + 1
        else:
            hi = mid
    
    return lo

def calc_multifactorial(n):
    val = 1;
    
    for i in range(9000, 1, -n):
        val *= i
        
    digit_cnt = 0;
    
    while val > 0:
        digit_cnt += 1
        val = val // 10
        
    return digit_cnt

for i in range(1, 9000):
    multifactorial.append(calc_multifactorial(i));
  
multifactorial.reverse();

for t in range(1, T+1):
    D = int(input())
    ans = 8998 - lower_bound(D) + 2
    
    print("Case #", end="")
    print(t, end="")
    print(": ", end="")
    
    if (ans == 9000):
        print("...")
    else:
        print("IT'S OVER 9000", end="")
        for i in range(ans):
            print("!", end="")
        print()
